chat/src/main/java/com/github/twitch4j/chat/TwitchChat.java

```diff
@@ -357,6 +391,49 @@ public TwitchChat(EventManager eventManager, CredentialManager credentialManager
                 }
             }
         });
+
+        // Initialize joinAttemptsByChannelName (on an attempt expiring without explicit removal, we retry with exponential backoff)
+        if (maxJoinRetries > 0) {
+            final long initialWait = initialWaitOverwrite != null ? initialWaitOverwrite : Math.max(chatQueueTimeout, 5000L);
+            this.joinAttemptsByChannelName = Caffeine.newBuilder()
+                .expireAfterWrite(initialWait, TimeUnit.MILLISECONDS)
+                .scheduler(Scheduler.forScheduledExecutorService(taskExecutor)) // required for prompt removals on java 8
+                .<String, Integer>evictionListener((name, attempts, cause) -> {
+                    if (cause == RemovalCause.EXPIRED && name != null && attempts != null) {
+                        if (attempts < maxJoinRetries) {
+                            taskExecutor.schedule(() -> {
+                                if (currentChannels.contains(name)) {
+                                    issueJoin(name, attempts + 1);
+                                }
+                            }, initialWait * (1L << (attempts + 1)), TimeUnit.MILLISECONDS); // exponential backoff (pow2 optimization)
+                        } else if (removeChannelOnJoinFailure && removeCurrentChannel(name)) {
+                            eventManager.publish(new ChannelRemovedPostJoinFailureEvent(name, ChannelRemovedPostJoinFailureEvent.Reason.RETRIES_EXHAUSTED));
+                        } else {
+                            log.debug("Chat connection exhausted retries when attempting to join channel: {}", name);
```

Comment:
Should we log this as error? I feel like this information is important if the user doesn't handle a `ChannelRemovedPostJoinFailureEvent`.

Reply:
Yeah I meant to make this at least a `warn`